src/dparse/lexer.d

```diff
@@ -1537,6 +1537,23 @@ private pure nothrow @safe:
             index);
     }
 
+    bool lexNamedEntity()
+    in { assert (range.bytes[range.index] == '&'); }
+    do
+    {
+        Token t;
+        range.popFront();
+        lexIdentifier(t, true);
+        if (t.type != tok!"identifier" ||
+            (!range.empty && range.bytes[range.index] != ';'))
+        {
+            error("Error: invalid named character entity");
+            return false;
+        }
+        range.popFront();
```

Comment:
I believe this can RangeError if the tokens end with `&` `identifier`, because we consumed the ampersand and the identifier, but the if invalid check didn't trigger because range.empty was true

For the same effect but a more sturdy check I would rewrite the if to
```d
if (t.type != tok!"identifier" || range.empty || range.bytes[range.index] != ';')
```

Reply:
Fixed. Also added a test case for `"\&0`

Reply:
that test case wouldn't match anyway because it already fails on the identifier, make the test case use an identifier instead

Reply:
Done